src/components/camera/CameraStream.jsx

```diff
@@ -13,24 +13,111 @@ import "./CameraStream.css";
  * Takes:
  *    cameraTitle: the camera title,
  *    cameraName: the camera name,
- *    unloadCallback, a callback that is ran before the window is fully unloaded
- * Returns a JSON object with keys: window, canvas, context, aspectRatio
+ *    unloadCallback: a callback that is ran before the window is fully unloaded
+ *    video_width: the stream width
+ *    video_height: the stream height
+ * Returns an object with keys: window, canvas, context, aspectRatio
  */
-function createPopOutWindow(cameraTitle, cameraName, unloadCallback) {
+function createPopOutWindow(cameraTitle, cameraName, unloadCallback, video_width, video_height) {
   let newWindow = window.open("", "", "width=500,height=500");
+  let style = newWindow.document.createElement('style');
+  style.innerHTML = `
+    #ext-wrapper {
+      background-color: black;
+      width: 100%;
+      height: 100%;
+      font-family: Arial, Helvetica, sans-serif;
+      display: flex;
+      justify-content: center;
+    }
+
+    #ext-title {
+      z-index: 100;
+      position: absolute;
+      width: 100%;
+      font-size: 30px;
+      font-weight: bold;
+      text-align: center;
+      color: white;
+      padding-top: 5px;
+      padding-bottom: 5px;
+      background-color: #00000066;
+    }
+
+    #ext-fps {
+      z-index: 101;
+      position: absolute;
+      top: 5;
+      left: 5;
+      font-size: 15px;
+      font-weight: bold;
+      color: red;
+    }
+
+    #ext-download-wrapper {
+      z-index: 101;
+      position: absolute;
+      bottom: 5px;
+      right: 5px;
+    }
+
+    #ext-download-button {
+      cursor: pointer;
+    }
+
+    #ext-download-button:disabled {
+      cursor: not-allowed;
+    }
+
+    #ext-vid {
+      z-index: 1;
+      border: none;
+      display: block;
+      margin: auto 0;
+    }
+  `;
+  newWindow.document.head.appendChild(style);
+  let script = newWindow.document.createElement('script');
+  script.innerHTML = `
+    function download() {
+      let canvas = document.getElementById("ext-vid");
+
+      let time = new Date();
+      let timezoneOffset = time.getTimezoneOffset() * 60000;
+      let timeString = new Date(time - timezoneOffset).toISOString().replace(":", "_").substring(0, 19);
+
+      let link = document.createElement("a");
+      
+      let tempCanvas = document.createElement('canvas');
+      tempCanvas.width = ${video_width} // video.videoWidth;
+      tempCanvas.height = ${video_height} // video.videoHeight;
+
+      let tempContext = tempCanvas.getContext('2d');
+      tempContext.drawImage(canvas, 0, 0, tempCanvas.width, tempCanvas.height);
+
+      link.href = tempCanvas.toDataURL("image/jpeg", 1);
+      link.download = "${cameraTitle}-" + timeString + ".jpg";
+      document.body.appendChild(link);
+      link.click();
+      document.body.removeChild(link);
+    }
```

Comment:
I'm not sure if it's a good idea to have code specified in a string. We don't have syntax highlighting or autocomplete or typechecking, so it would be really easy to introduce bugs here that are only triggered when downloading a frame.

Can we move all this HTML/CSS/JS in the popout window to standalone files, i.e. `cam_popout.html`, `cam_popout.css`, and `cam_popout.js`? Then you can specify that URL as an arg to `window.open()`. Since the code needs to know the camera title and video width and whatnot, you can add those as GET parameters and access them from the JS.

This approach makes it easier to edit that code, and makes the setup a bit cleaner and easier to read.